### ISSUE TITLE
Refactoring of CellValues and Fields

### ISSUE THREAD

**Issue by fverdugo** [gridap/Gridap.jl#30]
Place to keep miscellaneous comments about the ongoing re-factoring of CellValues and Co.

The re-factoring is taking place in the branch  refactoring_cell_values in folders `src/REFACTORED` and `test/REFACTORED`

**Comment by fverdugo** [comment]
With commit 0a9acd5217b7f6abd741849f33fffe18b2e23749, a first version of the new design of CellValues and Fields is available (for the moment only in branch refactoring_cell_values, where the code downstream from this part is still commented).

Features of the new design:

- Introduction of `NumberKernel` and `ArrayKernel`. Two types that represent operations that return numbers and arrays respectively. These operations are allowed to have arbitrary number of arguments. E.g. `CellsumKernel` takes an array and returns another array resulting from summing up all the entries in one dimension. Other available kernels are e.g., `VarinnerKarnel`, `LinCombKernel`.

- Introduction of the function `apply`. It takes a `NumberKernel` or an `ArrayKernel` and an arbitrary number of `CellValues`. It lazily applies the given kernel to the given `CellValues`.  E.g., applying the `VarinnerKernel` to a `CellBasis` and a `CellField` will return a `CellMap` object representing an (unevaluated)  element vector.  In the arguments of the `apply ` function, we can mix e.g., `CellMaps` and `CellArray` in an (almost) arbitrary way. E.g., we can apply the `LinCombKernel` to a `CellBasis` and a `CellVector`, which will result in a `CellField` . Applying the `LinCombKernel` to a `CellMatrix` and a `CellVector` will return a `CellVector`. The function `apply` has several overloaded methods for optimization. E.g.,  applying a kernel to a set of `ConstantCellValues` will return also a `ConstantCellValues`. Applying it to a set of  `IndexCellValues` will return an `IndexCellValues`.

- Expressivity, extensibility and performance. By defining its own kernels and using the `apply` function the user   will rarely need to implement `CellValue` object for their particular needs. The kernels also allow the fuse very complex operations instead of having a very complex operation trees between `CellValues`, which will definitively improve performance and compilation times

**Comment by fverdugo** [comment]
Things to be done:

- [ ] `apply` function can take up to 6 arguments for the moment. This can be easily generalized by using generated functions.
- [ ] An internal julia error es appearing very often. I have included a set of hacks to circumvent this error, but this has to be fixed better. I think it is related with the fact that I am storing tuples of objects in the fields of structs... but not sure. It can be related with this issue in julia (https://github.com/JuliaLang/julia/issues/29468)

**Comment by fverdugo** [comment]
I have opened an issue in julia about the internal error (https://github.com/JuliaLang/julia/issues/32322)

**Comment by fverdugo** [comment]
The refactoring is already available in master since PR #31.
Closing issue

**Comment by fverdugo** [comment]
Note for developers (specialy for @santiagobadia ):

PR #31 has introduces backward incompatible changes into master.

This is an illustrative (but strictly speaking incomplete) list of the changes to be aware of:

* `VectorValue{D,T}`, `TensorValue{D,T,L}`, and `Point{D,T}` are now parametrized also by `T` the type of the stored entries. Before `T` was always `Float64`. So, if, e.g., we want to initialize an array of points we need also to specify `T`:  `zeros(Point{3,Int},10,3)`

*  The `celldata` function used to get the unique value in a `ConstantCellValue` is removed. Use the field `value` instead.

* To get access to the cell value wrappers (e.g., `CellValueFromArray`) one has to write `using Gridap.CellValuesGallery`

* `celltypes` and `cellorders` functions have been removed from the `Triangulation` interface (now, they are only in the `Grid` interface). Instead, we have defined `CellRefFEs(::Triangulation)` that returns a `CellValue{<:RefFE}` (usually a `ConstantCellValue`) indicating the reference elements in the cells of the triangulation.

* Renamed `cellcoordinates(::Triangulation)` -> `CellPoints(::Triangulation)`

* Renamed `cellbasis(::Triangulation)` -> `CellBasis(::Triangulation)`

* Renamed `geomap(::Triangulation)` -> `CellGeomap(::Triangulation)`
